### PR TITLE
Caching of results using mongoDB. 

### LINKED ISSUE #380: Support caching of results
**I'm submitting a ...**
- [ ] bug report
- [x] feature request

Hey, currently, whenever a request comes in, query-server goes to the search engine every time to fetch the result. We can improve this, two ways
- Use something in python for caching
- Use mongodb to save results

I personally don't feel first option is good since, we'll need the results for quite a long time, so probably a lot of ram will be consumed with simple chaching, so I recommend mongodb.

There is already a basic [implementation](https://github.com/fossasia/query-server/blob/master/app/server.py#L65-L67)

### PR DESCRIPTION
**Pull request by starlord1311**
Fixes #380 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `master` branch.
- [x] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:

-made changes to server.py 
added db.save() so as the results of the query are saved in case they are not found in the db. so that they can be used next time.
-

### CODE REVIEW

**Comment by starlord1311** [comment]
@AnshulMalik @vaibhavsingh97 @raju249 @cclauss anything else required?

**Comment by cclauss** [comment]
My take on this one is that we need to:
* When someone makes a new query then we should see if that URL is already in the database.
* If it is and the timestamp is < one week old we should return the links from the DB
* If it is not then we should store a timestamp, the URL and the links that come back from every query.

**Comment by starlord1311** [comment]
@cclauss what changes should i make?

**Comment by starlord1311** [comment]
@cclauss the change i have made saves the query in the DB right?(without the timestamp). basically t does the 3rd step among the steps you told right?

**Comment by cclauss** [comment]
Think of the URL as the question and the links as the answer.  You need to save BOTH.

File: app/server.py [hunk 1]
```diff
@@ -63,6 +63,8 @@ def search(search_engine):
         if db['queries'].find({query: query}).limit(1) is False:
             db['queries'].insert(
                 {"query": query, "engine": engine, "qformat": qformat})
+            db['queries'].save(
+                {"query": query, "engine": engine, "qformat": qformat})
```
Comment (cclauss) [inline app/server.py]: My sense is that all three of these elements are contained in the URL.  Why not just store the URL?  We also need to store the links that come back from the query plus a timestamp.

File: app/server.py [hunk 2]
```diff
@@ -63,6 +63,8 @@ def search(search_engine):
         if db['queries'].find({query: query}).limit(1) is False:
             db['queries'].insert(
                 {"query": query, "engine": engine, "qformat": qformat})
+            db['queries'].save(
```
Comment (raju249) [inline app/server.py]: Why do we need both `save` and `insert` ?

File: app/server.py (hunk 1 above)
Comment (AnshulMalik) [reply app/server.py]: One more thing, the url contains the number of results to be fetched, this is one extra thing, but we can work on that later, let's not make this PR too complex

File: app/server.py [hunk 3]
```diff
@@ -63,6 +63,8 @@ def search(search_engine):
         if db['queries'].find({query: query}).limit(1) is False:
```
Comment (AnshulMalik) [inline app/server.py]: Need a change here, to consider not only query, but the engine as well.

File: app/server.py [hunk 4]
```diff
@@ -63,6 +63,8 @@ def search(search_engine):
...
             db['queries'].insert(
                 {"query": query, "engine": engine, "qformat": qformat})
```
Comment (AnshulMalik) [inline app/server.py]: And also, we need to check for tiimestamp as @cclauss said.
We also need to remove/update the outdated results.

Regarding the format of data, I think we can just store data in json in mongodb, and convert it ondemand, as requested in the request.

File: app/server.py (hunk 3 above)
Comment (bhaveshAn) [reply app/server.py]: Agree 👍

**Comment by cclauss** [comment]
Closed by #399 ?

**Pull request closed without merging**